File: scc.go
package main

import (
	"fmt"
	"os"
	"sort"
)
type Cycle_struct struct {
	member[] int
}
type Cycles []Cycle_struct
var cycle[] Cycle_struct
var ar[10001][] int
var rar[10001][] int
var visit[10001] bool
var stk[10001] int
var n,m,stk_size,cycle_cnt int
func init_visit() {
	for i:=1 ; i<=n ; i++ {
		visit[i]=false
	}
}
func push(val int) {
	stk[stk_size]=val
	stk_size++
}
func pop() int {
	stk_size--
	return stk[stk_size]
}
func dfs(x int) {
	visit[x]=true
	for _, next:=range ar[x] {
		if(!visit[next]) {
			dfs(next)
		}
	}
	push(x)
}
func rdfs(x int) {
	visit[x]=true
	cycle[cycle_cnt].member=append(cycle[cycle_cnt].member,x)
	for _, next:=range rar[x] {
		if(!visit[next]) {
			rdfs(next)
		}
	}
}
func scc() {
	init_visit()
	for i:=1 ; i<=n ; i++ {
		if(!visit[i]) {
			dfs(i)
		}
	}
	init_visit()
	for stk_size>0 {
		next:=pop()
		if(!visit[next]) {
			cycle=append(cycle,Cycle_struct{})
			rdfs(next)
			cycle_cnt++
		}
	}
}
func(s Cycles) Len() int {
	return cycle_cnt
}
func(s Cycles) Swap(i,j int) {
	s[i],s[j]=s[j],s[i]
}
func(s Cycles) Less(i,j int) bool {
	return s[i].member[0]<s[j].member[0]
}
func main() {
	in, _ := os.Open("input.txt")
	defer in.Close()
	out, _ := os.Create("output.txt")
	defer out.Close()
	fmt.Fscan(in,&n,&m)
	for i :=0; i<m ; i++ {
		var a,b int
		fmt.Fscan(in,&a,&b)
		ar[a]=append(ar[a],b)
		rar[b]=append(rar[b],a)
	}
	scc()
	for i:=0 ; i<cycle_cnt ; i++ {
		sort.Sort(sort.IntSlice(cycle[i].member))
	}
	sort.Sort(Cycles(cycle))
	fmt.Fprintln(out,"number of cycles : ",cycle_cnt)
	fmt.Println("number of cycles :",cycle_cnt)
	for i:=0 ; i<cycle_cnt ; i++ {
		fmt.Fprint(out,i+1,"th members : ")
		fmt.Print(i+1,"th members : ")
		for _, next:=range cycle[i].member {
			fmt.Fprint(out,next," ")
			fmt.Print(next," ")
		}
		fmt.Fprintln(out)
		fmt.Println()
	}
	fmt.Fprintln(out,"2013210109 Kim Hyung Jun")
	fmt.Println("2013210109 Kim Hyung Jun")
}

File: scc.lua
function scannum()
	local get=false
	local num=0
	while datahandle<=datalen do
		local next=tonumber(string.sub(data,datahandle,datahandle))
		if next and next>=0 and next<=9 then
			num=num*10+next
			get=true
		elseif get then
			return num
		end
		datahandle=datahandle+1
	end
	if get then return num end
	return nil
end

function init_visit()
	for i=1,n do
		visit[i]=false
	end
end

function push(x)
stk_size=stk_size+1
stk[stk_size]=x;
end

function pop()
local val=stk[stk_size]
stk[stk_size]=nil
stk_size=stk_size-1
return val
end

function dfs(x)
	visit[x]=true
	for _,next in pairs(ar[x]) do
		if visit[next]==false then
			dfs(next)
		end
	end
	push(x)
end

function rdfs(x)
	visit[x]=true
	cycle_size[cycle_cnt]=cycle_size[cycle_cnt]+1
	cycle[cycle_cnt][cycle_size[cycle_cnt]]=x
	for _,next in pairs(rar[x]) do
		if visit[next]==false then
			rdfs(next)
		end
	end
end

function scc()
	visit={}
	stk={}
	stk_size=0
	init_visit()
	for i=1,n do
		if visit[i]==false then
			dfs(i)
		end
	end
	init_visit()
	cycle={}
	cycle_cnt=0
	cycle_size={}
	while(stk_size>0) do
		next=pop()
		if visit[next]==false then
			cycle_cnt=cycle_cnt+1
			cycle[cycle_cnt]={}
			cycle_size[cycle_cnt]=0
			rdfs(next)
		end
	end
end
function cmp(i,j)
	return cycle[i][1]<cycle[j][1]
end
local input=io.open("input.txt","r")
local output=io.open("output.txt","w")
if input then
	data=input:read("*all")
	input:close()
end
datalen=string.len(data)
datahandle=1
n=scannum()
m=scannum()
ar={}
ar_size={}
rar={}
rar_size={}
for i=1,n do
	ar[i]={}
	ar_size[i]=0
	rar[i]={}
	rar_size[i]=0
end
for i=1,m do
	local a=scannum()
	local b=scannum()
	ar_size[a]=ar_size[a]+1
	ar[a][ar_size[a]]=b
	rar_size[b]=rar_size[b]+1
	rar[b][rar_size[b]]=a
end
scc()
order={}
for i=1,cycle_cnt do
	table.sort(cycle[i])
	order[i]=i
end
table.sort(order,cmp)

output:write("number of cycles : ",cycle_cnt,'\n')
for i=1,cycle_cnt do
	output:write(i,"th members : ")
	for _,member in pairs(cycle[order[i]]) do
		output:write(member,' ')
	end
	output:write('\n')
end
output:write("2013210109 Kim Hyung Jun\n")
output:close()
